purchase_allowed_product/models/product.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# © 2017 Today Mourad EL HADJ MIMOUNE @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    @api.model
+    def search(self, args, offset=0, limit=None, order=None, count=False):
+        restrict_supplier_id = self.env.context.get(
+            'restrict_supplier_id', False)
+        use_only_supplied_product = self.env.context.get(
+            'use_only_supplied_product', False)
+        if use_only_supplied_product:
+            seller = self.env['res.partner'].browse(restrict_supplier_id)
+            seller = seller.commercial_partner_id if seller.\
+                commercial_partner_id else seller
+            supplierinfos = self.env['product.supplierinfo'].search(
+                [('name', '=', seller.id)])
+            args.append(
+                ('product_tmpl_id', 'in',
+                    [x.product_tmpl_id.id for x in supplierinfos]))
```

Comment:
ok but as seen with @sebastienbeau there is a problem because in product.supplierinfo there are product_id and product_tmpl_id field. In your search there is only search on product_tmpl_id. The model changed in v10